mcts_new.py

```diff
@@ -0,0 +1,280 @@
+        visit_probabilities = visit_counts ** (1 / temperature)
+        visit_probabilities = visit_probabilities / visit_probabilities.sum()
+        return visit_probabilities
+
+
+@dataclasses.dataclass(slots=True)
+class AfterStateNode:
+    state: sj.Skyjo
+    action: sj.SkyjoAction
+    parent: DecisionStateNode
+    children: dict[sj.Skyjo, DecisionStateNode | TerminalStateNode] = dataclasses.field(
+        default_factory=dict
+    )
+    realized_counts: dict[sj.Skyjo, int] = dataclasses.field(default_factory=dict)
+    num_visits: int = 0
+    average_visit_value: float = 1
+    is_expanded: bool = False
+
+    def _realize_outcome(self) -> sj.Skyjo:
+        outcome_state = sj.apply_action(self.state, self.action)
+        outcome_state_hash = sj.hash_skyjo(outcome_state)
+        self.realized_counts[outcome_state_hash] = (
+            self.realized_counts.get(outcome_state_hash, 0) + 1
+        )
+        return outcome_state
+
+    def expand(self, model: skynet.SkyNet):
+        # Placeholder for now we may want to simulate many random outcomes here
+        self.is_expanded = True
+
+    def select_child(
+        self, model: skynet.SkyNet
+    ) -> DecisionStateNode | TerminalStateNode:
+        """Realize next state by applying action. Returns node in game tree that represents realized next state."""
+        realized_next_state = self._realize_outcome()
+        if sj.get_game_over(realized_next_state):
+            return TerminalStateNode(
+                state=realized_next_state,
+                parent=self,
+                prev_action=self.action,
+            )
+        realized_next_state_hash = sj.hash_skyjo(realized_next_state)
+        if realized_next_state_hash not in self.children:
+            self.children[realized_next_state_hash] = DecisionStateNode(
+                state=realized_next_state,
+                parent=self,
+                prev_action=self.action,
+                model_output=model.predict(realized_next_state),
+            )
+
+        return self.children[realized_next_state_hash]
+
+    def realized_outcome_probability(self, realized_outcome_state: sj.Skyjo) -> float:
+        realized_outcome_state_hash = sj.hash_skyjo(realized_outcome_state)
+
+        return self.realized_counts.get(realized_outcome_state_hash, 0) / sum(
+            self.realized_counts.values()
+        )
+
+
+@dataclasses.dataclass(slots=True)
+class TerminalStateNode:
+    state: sj.Skyjo
+    parent: AfterStateNode | DecisionStateNode
+    prev_action: sj.SkyjoAction
+    num_visits: int = 0
+    average_visit_value: float = 1
+    is_expanded: bool = False
+
+    @property
+    def outcome(self) -> skynet.StateValue:
+        return skynet.skyjo_to_state_value(self.state)
+
+    def expand(self, model: skynet.SkyNet):
+        raise ValueError("Terminal nodes should not need to be expanded")
+
+
+MCTSNode = DecisionStateNode | AfterStateNode | TerminalStateNode
```

Comment:
```suggestion
MCTSNode: TypeAlias = DecisionStateNode | AfterStateNode | TerminalStateNode
```

plus `from typing import TypeAlias`

plays better with mypy if we ever enable that